pkg/util/metric/metric.go

```diff
@@ -176,10 +186,261 @@ func maybeTick(m periodic) {
+	}
+}
+
+// NewHdrHistogram initializes a given Histogram. The contained windowed histogram
+// rotates every 'duration'; both the windowed and the cumulative histogram
+// track nonnegative values up to 'maxVal' with 'sigFigs' decimal points of
+// precision.
+func NewHdrHistogram(
+	metadata Metadata, duration time.Duration, maxVal int64, sigFigs int,
+) *HdrHistogram {
+	h := &HdrHistogram{
+		Metadata: metadata,
+		maxVal:   maxVal,
+	}
+	wHist := hdrhistogram.NewWindowed(histWrapNum, 0, maxVal, sigFigs)
+	h.mu.cumulative = hdrhistogram.New(0, maxVal, sigFigs)
+	h.mu.sliding = wHist
+	h.mu.tickHelper = &tickHelper{
+		nextT:        now(),
+		tickInterval: duration / histWrapNum,
+		onTick: func() {
+			wHist.Rotate()
+		},
+	}
+	return h
+}
+
+// NewHdrLatency is a convenience function which returns a histogram with
+// suitable defaults for latency tracking. Values are expressed in ns,
+// are truncated into the interval [0, MaxLatency] and are recorded
+// with one digit of precision (i.e. errors of <10ms at 100ms, <6s at 60s).
+//
+// The windowed portion of the Histogram retains values for approximately
+// histogramWindow.
+func NewHdrLatency(metadata Metadata, histogramWindow time.Duration) *HdrHistogram {
+	return NewHdrHistogram(
+		metadata, histogramWindow, MaxLatency.Nanoseconds(), 1,
+	)
+}
+
+// RecordValue adds the given value to the histogram. Recording a value in
+// excess of the configured maximum value for that histogram results in
+// recording the maximum value instead.
+func (h *HdrHistogram) RecordValue(v int64) {
+	h.mu.Lock()
+	defer h.mu.Unlock()
+
+	if h.mu.sliding.Current.RecordValue(v) != nil {
+		_ = h.mu.sliding.Current.RecordValue(h.maxVal)
+	}
+	if h.mu.cumulative.RecordValue(v) != nil {
+		_ = h.mu.cumulative.RecordValue(h.maxVal)
+	}
+}
+
+// TotalCount returns the (cumulative) number of samples.
+func (h *HdrHistogram) TotalCount() int64 {
+	h.mu.Lock()
+	defer h.mu.Unlock()
+	return h.mu.cumulative.TotalCount()
+}
+
+// Min returns the minimum.
+func (h *HdrHistogram) Min() int64 {
+	h.mu.Lock()
+	defer h.mu.Unlock()
+	return h.mu.cumulative.Min()
+}
+
+// Inspect calls the closure with the empty string and the receiver.
+func (h *HdrHistogram) Inspect(f func(interface{})) {
+	h.mu.Lock()
+	maybeTick(h.mu.tickHelper)
+	h.mu.Unlock()
+	f(h)
+}
+
+// GetType returns the prometheus type enum for this metric.
+func (h *HdrHistogram) GetType() *prometheusgo.MetricType {
+	return prometheusgo.MetricType_HISTOGRAM.Enum()
+}
+
+// ToPrometheusMetric returns a filled-in prometheus metric of the right type.
+func (h *HdrHistogram) ToPrometheusMetric() *prometheusgo.Metric {
+	hist := &prometheusgo.Histogram{}
+
+	h.mu.Lock()
+	maybeTick(h.mu.tickHelper)
+	bars := h.mu.cumulative.Distribution()
+	hist.Bucket = make([]*prometheusgo.Bucket, 0, len(bars))
+
+	var cumCount uint64
+	var sum float64
+	for _, bar := range bars {
+		if bar.Count == 0 {
+			// No need to expose trivial buckets.
```

Comment:
Alert, this brings a bug back. We do absolutely need to expose trivial buckets. Where did you get this code from? Are there any other changes we're accidentally reverting?

https://github.com/cockroachdb/cockroach/pull/88331/files#diff-7d9143ffc52f8bb21b9a9f1dfdcef120129a14453e54e126e9ba6e3da4b54941R297-R298

Reply:
Ah, truly between a rock and a hard place here. Got it, let's use the broken histograms then..

Reply:
Ack - thanks for chiming in @dhartunian. Leaving this as-is, then.